tests/Microsoft.Identity.Test.Unit.net45/Throttling/ThrottlingAcceptanceTests.cs

```diff
@@ -233,10 +250,113 @@ await app.AcquireTokenSilent(new[] { "Other.Scopes" }, account).ExecuteAsync()
+        #region UiRequired test
+
+
+        /// <summary>
+        /// UI required cache for the same request (invalid_grant)
+        /// Create a request that should have the response with HTTP status 400 and OAuth error invalid_grant.
+        /// Repeat the same request immediately - expectation is the result should be returned from UI required cache.
+        /// Sleep for at least DefaultUIRequired + 1 seconds.
+        /// Repeat the same request again - expectation is the request should be issued on the server.
+        /// If a request had the response with Retry After header, then a different request with a different strict thumbprint should work
+        /// </summary>
+        [TestMethod]
+        public async Task UiRequiredThrottling_Async()
+        {
+            using (var httpManagerAndBundle = new MockHttpAndServiceBundle())
+            {
+                var app = await SetupAndAquireOnceAsync(
+                    httpManagerAndBundle, 
+                    400, 
+                    null, 
+                    TokenResponseType.InvalidGrant).ConfigureAwait(false);
+
+                var account = (await app.GetAccountsAsync().ConfigureAwait(false)).Single();
+                var throttlingManager = (httpManagerAndBundle.ServiceBundle.ThrottlingManager as SingletonThrottlingManager);
+                AssertThrottlingCacheEntryCount(throttlingManager, uiRequiredEntryCount: 1);
+
+                Trace.WriteLine("A similar request will be throttled");
+                var ex = await AssertException.TaskThrowsAsync<MsalThrottledUiRequiredException>(
+                   () => app.AcquireTokenSilent(TestConstants.s_scope, account)
+                        .ExecuteAsync())
+                       .ConfigureAwait(false);
+                Assert.AreEqual(400, ex.StatusCode);
+                AssertThrottlingCacheEntryCount(throttlingManager, uiRequiredEntryCount: 1);
+
+                Trace.WriteLine("And again...");
+
+                ex = await AssertException.TaskThrowsAsync<MsalThrottledUiRequiredException>(
+                  () => app.AcquireTokenSilent(TestConstants.s_scope, account)
+                       .ExecuteAsync())
+                      .ConfigureAwait(false);
+                Assert.AreEqual(400, ex.StatusCode);
+                AssertThrottlingCacheEntryCount(throttlingManager, uiRequiredEntryCount: 1);
+
+                Trace.WriteLine("Time passes, the same request should now pass");
+                throttlingManager.SimulateTimePassing(
+                    UiRequiredProvider.s_uiRequiredExpiration + 
+                    TimeSpan.FromSeconds(1));
+
+                httpManagerAndBundle.HttpManager.AddTokenResponse(TokenResponseType.Valid);
+                await app.AcquireTokenSilent(TestConstants.s_scope, account).ExecuteAsync()
+                       .ConfigureAwait(false);
+
+                AssertThrottlingCacheEntryCount(throttlingManager, uiRequiredEntryCount: 0);
+            }
```

Comment:
can the uiRequiredEntryCount ever get above 1?

Reply:
Yes. I will add a test to show this. The entries in the dictionary depend on: 

- client ID
- authority
- refresh token
- scopes

So if you make 2 requests that result in a UIRequired exception, and the 2 requests differ because of any of the above params, there will be 2 entries. 

Added a test where we request fewer scopes - it shows 2 entries.